pkg/controller/certificates/authority/authority.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authority
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/x509"
+	"fmt"
+	"math/big"
+	"time"
+)
+
+var serialNumberLimit = new(big.Int).Lsh(big.NewInt(1), 128)
+
+type CertificateAuthority struct {
+	Certificate *x509.Certificate
+	PrivateKey  crypto.Signer
+	Backdate    time.Duration
+	Now         func() time.Time
+}
+
+func (ca *CertificateAuthority) Sign(crDER []byte, policy SigningPolicy) ([]byte, error) {
+	now := time.Now()
+	if ca.Now != nil {
+		now = ca.Now()
+	}
+
+	nbf := now.Add(-ca.Backdate)
+	if ca.Certificate.NotAfter.Before(nbf) {
+		return nil, fmt.Errorf("the signer has expired: NotAfter=%v", ca.Certificate.NotAfter)
+	}
+
+	cr, err := x509.ParseCertificateRequest(crDER)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse certificate request: %v", err)
+	}
+	if err := cr.CheckSignature(); err != nil {
+		return nil, fmt.Errorf("unable to verify certificate request signature: %v", err)
+	}
+
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate a serial number for %s: %v", cr.Subject.CommonName, err)
+	}
+
+	tmpl := &x509.Certificate{
+		SerialNumber:       serialNumber,
+		Subject:            cr.Subject,
+		DNSNames:           cr.DNSNames,
+		IPAddresses:        cr.IPAddresses,
+		EmailAddresses:     cr.EmailAddresses,
+		URIs:               cr.URIs,
```

Comment:
if we're adding support for URIs, also update describer:

https://github.com/kubernetes/kubernetes/blob/7b6369c803cd06f0084ac57b2b6266000b739aea/staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go#L3264-L3273

and kubelet client cert recognizers to require URI SANs to be empty:

https://github.com/kubernetes/kubernetes/blob/7b6369c803cd06f0084ac57b2b6266000b739aea/pkg/controller/certificates/approver/sarapprove.go#L176-L178